### PR TITLE
Night vision goggles now have night vision.

### PR DESCRIPTION
**Pull request by MaxIsJoe**
### Purpose
Adds functionality to night vision goggles. They will now allow players to see a bit clearer in the dark.

image

### Changelog:

CL: Created a Night Vision Goggle script that enables the night vision shader and makes player vision a bit clearer in the dark.
- Added a reference to the Minimal Visibility Sprite game object in CameraEffectsControl.cs
- `AdjustPlayerVisibility(Vector3 newSize)` function has been added to adjust the player's vision in the dark.

### CODE REVIEW

File: UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+			camera.GetComponent<CameraEffectControlScript>().AdjustPlayerVisibility(defaultVisionVisibility);
+		}
+
+	private void OnWearing()
+		{
+			var camera = Camera.main;
+			camera.GetComponent<CameraEffectControlScript>().ToggleNightVisionEffectState();
+			camera.GetComponent<CameraEffectControlScript>().AdjustPlayerVisibility(nightVisionVisibility);
+		}
+}
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs]: This is serverside only, and wont affect clients. You'll need to make a TargetRPC

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs]: just put `[TargetRpc]` above the function, right?

Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs]: Add it above OnTakingOff and OnWearing, you'll also need to rename them to TargetOnTakingOff  and TargetOnWearing

Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs]: They'll also need this ```TargetOnTakingOff(NetworkConnection target)``` parameter added

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs]: do i need to change anything inside the function as well?

File: UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	[SerializeField, Tooltip("How far the player will be able to see in the dark while he has the goggles on.")]
+	private Vector3 nightVisionVisibility;
+
+	[SerializeField, Tooltip("The default minimal visibility size.")]
+	private Vector3 defaultVisionVisibility;
+
+
+	public void OnInventoryMoveServer(InventoryMove info)
+		{
+			RegisterPlayer registerPlayer;
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs]: RegisterPlayer registerPlayer = info.ToRootPlayer;

Then you can get rid of the ones at line 21 and 31

File: UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs (hunk 1 above)
Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs]: nope

**Comment by MaxIsJoe** [comment]
Unity reports "NetworkConnection, TargetRPC, TargetRPCAttribute cannot be found". 

What did we miss when making the changes?

File: UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	public void OnInventoryMoveServer(InventoryMove info)
+		{
+			RegisterPlayer registerPlayer = info.ToRootPlayer;
+
+			if (info.ToSlot != null && info.ToSlot?.NamedSlot != null)
+			{
+
+				if (registerPlayer != null && info.ToSlot.NamedSlot == NamedSlot.eyes)
+				{
+					TargetOnWearing();
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs]: you'll need to add the parameter, here. not sure if info.conn is a thing

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs]: still need to solve this issue first https://github.com/unitystation/unitystation/pull/6112#issuecomment-785094553

Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs]: change the class from a monobehaviour to NetworkBehaviour

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs]: still nothing

Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs]: using Mirror

File: UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using CameraEffects;
+
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs:6]: using Mirror; here

File: UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs (hunk 3 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs]: image

File: UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs [hunk 5]
```diff
@@ -1,57 +1,46 @@
...
 			{
 
 				if (registerPlayer != null && info.ToSlot.NamedSlot == NamedSlot.eyes)
 				{
-					TargetOnWearing(registerPlayer.connectionToClient);
+					OnWearing();
 				}
 			}
 
 			if (info.FromSlot != null && info.FromSlot?.NamedSlot != null && info.ToSlot != null)
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs]: if (info.FromSlot != null && info.FromSlot?.NamedSlot != null) for line 29

Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs]: as info.ToSlot != null is allowed to be null if we are removing it from the inventory directly

**Comment by corp-0** [comment]
Last I knew this is still wip, added DNM until it is tested and ready to merge.

**Comment by MaxIsJoe** [comment]
> Last I knew this is still wip, added DNM until it is tested and ready to merge.

Like I said on discord, I'm waiting for Dan to fix connection problem. If he can't find a way to fix this then we'll revert back to 03ee9f19c38233d7f49701b5642e2b04ce6b4bcb and work on the `ClientInventoryMove` improvements in a separate pull request.

And who knows? maybe the issue is from another thing that's not our problem and it will get fixed before tomorrow.

**Comment by corp-0** [comment]
Yeah, the label is just prevempting someone from accidentaly merging thinking it was ready.

**Comment by ThatDan123** [comment]
Im gonna be a bit busy over the next few days, so revert to the working commit for now. And put a note in the script that it'll need syncing to rejoinning clients at some point.

**Comment by MaxIsJoe** [comment]
The revert has been tested and it works fine, I think we can finally have one last review and then merge this.

**Comment by DTraitor** [comment]
> The revert has been tested and it works fine, I think we can finally have one last review and then merge this.

Test if night vision works after relog. Also how good it works with shuttles?

**Comment by MaxIsJoe** [comment]
> > The revert has been tested and it works fine, I think we can finally have one last review and then merge this.
> 
> Test if night vision works after relog. Also how good it works with shuttles?

How do I test that in playmode inside the editor? and wdym?

**Comment by DTraitor** [comment]
My bad, I forgot that you cant join as client in editor.
> and wdym?

destroy all the light on evac shuttle and test if night vision works correct while shuttle is moving

**Comment by MaxIsJoe** [comment]
> My bad, I forgot that you cant join as client in editor.
> 
> > and wdym?
> 
> destroy all the light on evac shuttle and test if night vision works correct while shuttle is moving

No issues as far as I can tell from my testing. I guess we can merge now.

File: UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+		if(check == true)
+		{
+			camera.GetComponent<CameraEffectControlScript>().ToggleNightVisionEffectState();
+			camera.GetComponent<CameraEffectControlScript>().AdjustPlayerVisibility(nightVisionVisibility);
+		}
+		else
+		{
+			camera.GetComponent<CameraEffectControlScript>().ToggleNightVisionEffectState();
+			camera.GetComponent<CameraEffectControlScript>().AdjustPlayerVisibility(defaultVisionVisibility);
+		}
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs]: Can you remove these indents

File: UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+	
+	private void OnStartClient()
+	{
+		if(PlayerManager.LocalPlayerScript != null)
+		{
+			var item = PlayerManager.LocalPlayerScript.Equipment.GetClothingItem(NamedSlot.eyes).GameObjectReference;
+			if(item == gameObject) {enableEffect(true);}
+		}
+	}
```
Comment (ThatDan123) [inline UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs]: Also add a note here saying this wont work as equipment wont be sync'd in time for the check, and that IServerInventoryMove should be replaced with IClienInventoryMove but that needs more functionality first

File: UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs (hunk 6 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs]: I can't find them, vscode is also starting to annoy me about this problem.

Comment (ThatDan123) [reply UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs]: i did it for you, strange the vscode wasnt showing them

**Review by ThatDan123 — APPROVED** [review]

File: UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+using Mirror;
+using CameraEffects;
+
+public class NightVisionGoggles : NetworkBehaviour, IServerInventoryMove
+{
+	[SerializeField, Tooltip("How far the player will be able to see in the dark while he has the goggles on.")]
+	private Vector3 nightVisionVisibility;
+
+	[SerializeField, Tooltip("The default minimal visibility size.")]
+	private Vector3 defaultVisionVisibility;
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Clothing/Eyewear/NightVisionGoggles.cs:13]: in the future this should read directly what the creature has as default visibility, I think. So we have creatures than can see in the dark without goggles.

**Pull request merged**